README.md
# Terra

**React geospatial visualization library** built on Deck.gl and MapLibre GL. Pre-built layers for heatmaps, point clusters, animated trajectories, geofence zones, and icon markers with a clean, composable API.

> Think of it as a higher-level `react-map-gl` with batteries included.

```
[ Screenshot: Dark-themed map with a plasma heatmap overlay showing earthquake   ]
[ density across the Mediterranean. Cluster markers visible at lower zoom levels ]
[ with animated vehicle trajectories rendered as glowing green trails.            ]
```

---

## Features

- **HeatmapLayer** -- Weighted density heatmaps with configurable color scales (heat, viridis, plasma, coolwarm)
- **ClusterLayer** -- Automatic spatial clustering with zoom-based expansion and click-to-zoom
- **TrajectoryLayer** -- Animated path rendering with trails, directional arrows, and endpoint markers
- **GeofenceLayer** -- Polygon zones with real-time enter/exit detection for tracked points
- **IconLayer** -- Custom markers (emoji, SVG, image) with hover tooltips and click handlers
- **Built-in controls** -- Zoom, layer toggle panel, geocoding search box
- **Hooks** -- `useViewport`, `useGeoJSON`, `useRealtime` (WebSocket subscriptions)
- **Geo utilities** -- Haversine distance, point-in-polygon, bounding box, path simplification
- **Color scales** -- Interpolation, sampling, domain mapping with built-in presets
- **Themes** -- Dark, satellite, light, and streets base maps out of the box

---

## Install

```bash
npm install @terra-gl/react
```

### Peer dependencies

```bash
npm install react react-dom @deck.gl/core @deck.gl/layers @deck.gl/aggregation-layers @deck.gl/react maplibre-gl
```

---

## Quickstart

```tsx
import {
  TerraMap,
  HeatmapLayer,
  TrajectoryLayer,
  GeofenceLayer,
  ZoomControl,
} from '@terra-gl/react';

function App() {
  return (
    <TerraMap center={[35.5, 33.9]} zoom={8} theme="dark">
      <HeatmapLayer data={earthquakes} weight="magnitude" radius={30} />
      <TrajectoryLayer data={vehicleTracks} color="#00ff88" animated />
      <GeofenceLayer
        zones={restrictedAreas}
        onEnter={(e) => alert(`Entered ${e.zone.name}`)}
      />
      <ZoomControl position="top-right" />
    </TerraMap>
  );
}
```

```
[ Screenshot: The quickstart example rendered -- a dark map centered on Beirut  ]
[ with earthquake heatmap, animated vehicle paths, and red geofence polygons.   ]
```

---

## API Reference

### `<TerraMap>`

Root container component. All layers and controls must be children of `TerraMap`.

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `center` | `[lon, lat]` | `[0, 0]` | Initial map center |
| `zoom` | `number` | `2` | Initial zoom level (0-22) |
| `pitch` | `number` | `0` | Map pitch in degrees |
| `bearing` | `number` | `0` | Map bearing in degrees |
| `theme` | `'dark' \| 'satellite' \| 'light' \| 'streets' \| string` | `'dark'` | Theme or custom style URL |
| `width` | `string \| number` | `'100%'` | Container width |
| `height` | `string \| number` | `'100%'` | Container height |
| `interactive` | `boolean` | `true` | Enable/disable user interactions |
| `onViewportChange` | `(viewport) => void` | -- | Viewport change callback |
| `onClick` | `(info) => void` | -- | Click callback with pick info |

The component exposes an imperative handle via `React.forwardRef`:

```tsx
const mapRef = useRef<TerraMapHandle>(null);

mapRef.current.flyTo(-73.98, 40.75, 12);
mapRef.current.fitBounds([-74.1, 40.6, -73.8, 40.9]);
```

---

### Layers

#### `<HeatmapLayer>`

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `data` | `T[]` | required | Point data array |
| `weight` | `keyof T \| number` | `1` | Weight field name or constant |
| `radius` | `number` | `30` | Heatmap radius in pixels |
| `intensity` | `number` | `1` | Intensity multiplier |
| `colorScale` | `'heat' \| 'viridis' \| 'plasma' \| 'coolwarm' \| RGBA[]` | `'heat'` | Color scale |
| `opacity` | `number` | `0.8` | Layer opacity |
| `getPosition` | `(d: T) => [lon, lat]` | -- | Custom position accessor |
| `getWeight` | `(d: T) => number` | -- | Custom weight accessor |

#### `<ClusterLayer>`

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `data` | `T[]` | required | Point data array |
| `clusterRadius` | `number` | `50` | Clustering radius in pixels |
| `maxClusterZoom` | `number` | `14` | Max zoom for clustering |
| `minPoints` | `number` | `2` | Min points per cluster |
| `clusterColor` | `ColorValue` | `'#ff6b6b'` | Cluster circle color |
| `pointColor` | `ColorValue` | `'#4dabf7'` | Unclustered point color |
| `onClusterClick` | `(cluster) => void` | -- | Cluster click handler |
| `onPointClick` | `(point) => void` | -- | Point click handler |

#### `<TrajectoryLayer>`

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `data` | `Trajectory[]` | required | Trajectory data |
| `color` | `ColorValue` | `'#00ff88'` | Path color |
| `width` | `number` | `2` | Path width in pixels |
| `animated` | `boolean` | `false` | Enable path animation |
| `speed` | `number` | `1` | Animation speed multiplier |
| `trailLength` | `number` | `0.15` | Trail length (0-1) |
| `showArrows` | `boolean` | `false` | Show direction arrows |
| `showEndpoints` | `boolean` | `false` | Show start/end markers |

#### `<GeofenceLayer>`

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `zones` | `GeofenceZone[]` | required | Polygon zone definitions |
| `fillColor` | `ColorValue` | `'rgba(66,135,245,0.2)'` | Zone fill color |
| `strokeColor` | `ColorValue` | `'#4287f5'` | Zone border color |
| `showLabels` | `boolean` | `false` | Show zone name labels |
| `trackedPoints` | `Coordinate[]` | -- | Points to track for enter/exit |
| `onEnter` | `(event) => void` | -- | Zone enter callback |
| `onExit` | `(event) => void` | -- | Zone exit callback |

#### `<IconLayer>`

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `data` | `T[]` | required | Marker data array |
| `icon` | `string` | `'📍'` | Default icon (emoji/URL) |
| `iconSize` | `number` | `24` | Icon size in pixels |
| `showTooltip` | `boolean` | `false` | Enable hover tooltips |
| `renderTooltip` | `(item: T) => ReactNode` | -- | Custom tooltip renderer |
| `onClick` | `(item: T) => void` | -- | Click handler |

---

### Controls

#### `<ZoomControl>`

```tsx
<ZoomControl position="top-right" showCompass />
```

#### `<LayerToggle>`

```tsx
<LayerToggle
  layers={[
    { id: 'heatmap', label: 'Heatmap', visible: true },
    { id: 'clusters', label: 'Clusters', visible: false },
  ]}
  onToggle={(id, visible) => handleToggle(id, visible)}
/>
```

#### `<SearchBox>`

```tsx
<SearchBox
  position="top-left"
  onSelect={(result) => mapRef.current.flyTo(...result.position, 14)}
/>
```

---

### Hooks

#### `useViewport`

```tsx
const { viewport, flyTo, fitBounds, zoomIn, zoomOut } = useViewport({
  longitude: -73.98,
  latitude: 40.75,
  zoom: 12,
  minZoom: 4,
  maxZoom: 18,
});
```

#### `useGeoJSON`

```tsx
const { data, loading, error, refetch } = useGeoJSON({
  url: 'https://api.example.com/points.geojson',
  refetchInterval: 30000,
});
```

#### `useRealtime`

```tsx
const { data, status, send } = useRealtime({
  url: 'wss://api.example.com/vehicles/stream',
  messageFilter: 'position_update',
  reconnect: true,
});
```

---

### Utilities

#### Geo

```ts
import { haversineDistance, pointInPolygon, computeBBox, simplifyPath } from '@terra-gl/react';

const km = haversineDistance([35.5, 33.9], [35.8, 34.1]);
const inside = pointInPolygon([35.5, 33.9], polygonCoords);
const bbox = computeBBox(points);
const simplified = simplifyPath(densePath, 0.001);
```

#### Color

```ts
import { createColorScale, domainColor, parseColor } from '@terra-gl/react';

const scale = createColorScale('viridis');
const color = scale(0.5); // [38, 130, 142, 255]
const mapped = domainColor(6.5, [0, 10], scale);
```

---

## Themes

| Theme | Description |
|-------|-------------|
| `dark` | Dark Matter (Carto) -- ideal for data-dense overlays |
| `satellite` | Voyager (Carto) -- good for terrain context |
| `light` | Positron (Carto) -- clean, minimal background |
| `streets` | Voyager variant -- labeled roads and POIs |

Pass a custom MapLibre style URL for full control:

```tsx
<TerraMap theme="https://tiles.example.com/style.json" />
```

---

## Data Types

```ts
interface WeightedPoint {
  position: [longitude: number, latitude: number];
  weight?: number;
}

interface Trajectory {
  id: string;
  path: { position: [lon, lat]; timestamp: number }[];
}

interface GeofenceZone {
  id: string;
  name: string;
  polygon: [lon, lat][];
}

interface IconMarker {
  id: string;
  position: [lon, lat];
  icon?: string;
  label?: string;
}
```

---

## Development

```bash
git clone https://github.com/Waiel5/terra.git
cd terra
npm install
npm run dev          # watch mode
npm run storybook    # component stories
npm test             # run tests
```

---

## License

MIT


CHANGELOG.md
# Changelog
- 2021-04-11 06:04: update #1
- 2021-04-14 06:24: update #2
- 2021-04-18 04:00: update #3
- 2021-04-20 01:56: update #4
- 2021-04-20 06:18: update #5
- 2021-04-20 02:04: update #6
- 2021-04-22 05:01: update #7
- 2021-04-22 17:42: update #8
- 2021-04-23 19:27: update #9
- 2021-04-23 20:41: update #10
- 2021-04-24 19:42: update #11
- 2021-04-25 16:16: update #12
- 2021-04-27 23:02: update #13
- 2021-05-01 00:00: update #14
- 2021-05-02 06:44: update #15
- 2021-05-01 21:14: update #16
- 2021-05-03 23:02: update #17
- 2021-05-04 16:27: update #18
- 2021-05-05 05:20: update #19
- 2021-05-05 19:11: update #20
- 2021-05-07 05:25: update #21
- 2021-05-07 20:42: update #22
- 2021-05-07 21:22: update #23
- 2021-05-08 04:54: update #24
- 2021-05-10 00:03: update #25
- 2021-05-10 18:23: update #26
- 2021-05-10 23:10: update #27
- 2021-05-11 02:32: update #28
- 2021-05-10 18:28: update #29
- 2021-05-11 18:19: update #30
- 2021-05-12 18:55: update #31
- 2021-05-13 06:12: update #32
- 2021-05-12 18:34: update #33
- 2021-05-14 17:54: update #34
- 2021-05-16 19:14: update #35
- 2021-05-18 17:43: update #36
- 2021-05-19 00:11: update #37
- 2021-05-19 22:33: update #38
- 2021-05-21 17:42: update #39
- 2021-05-22 03:04: update #40
- 2021-05-25 05:57: update #41
- 2021-05-25 02:22: update #42
- 2021-05-27 23:26: update #43
- 2021-05-27 23:24: update #44
- 2021-05-29 02:15: update #45
- 2021-05-29 18:17: update #46
- 2021-05-31 04:52: update #47
- 2021-06-02 01:42: update #48
- 2021-06-02 01:42: update #49
- 2021-06-03 01:28: update #50
- 2021-06-05 05:46: update #51
- 2021-06-07 04:31: update #52
- 2021-06-07 21:13: update #53
- 2021-06-08 02:29: update #54
- 2021-06-09 04:21: update #55
- 2021-06-10 05:02: update #56
- 2021-06-10 05:32: update #57
- 2021-06-11 17:26: update #58
- 2021-06-13 06:06: update #59
- 2021-06-14 19:05: update #60
- 2021-06-17 06:39: update #61
- 2021-06-17 01:39: update #62
- 2021-06-18 20:58: update #63
- 2021-06-20 22:20: update #64
- 2021-06-22 06:43: update #65
- 2021-06-24 06:35: update #66
- 2021-06-25 22:02: update #67
- 2021-06-29 04:35: update #68
- 2021-06-29 06:13: update #69
- 2021-06-30 06:43: update #70
- 2021-06-30 22:23: update #71
- 2021-07-01 22:15: update #72
- 2021-07-03 00:17: update #73
- 2021-07-04 16:20: update #74
- 2021-07-06 05:36: update #75
- 2021-07-06 19:26: update #76
- 2021-07-07 18:08: update #77
- 2021-07-12 02:01: update #78
- 2021-07-13 06:03: update #79
- 2021-07-18 02:35: update #80
- 2021-07-18 23:46: update #81
- 2021-07-21 17:29: update #82
- 2021-07-22 18:30: update #83
- 2021-07-29 01:18: update #84
- 2021-07-30 19:32: update #85
- 2021-08-01 06:06: update #86
- 2021-08-03 03:36: update #87
- 2021-08-02 17:12: update #88
- 2021-08-03 04:52: update #89
- 2021-08-04 02:40: update #90
- 2021-08-05 23:31: update #91
- 2021-08-06 03:50: update #92
- 2021-08-07 00:29: update #93
- 2021-08-07 06:57: update #94
- 2021-08-10 01:00: update #95
- 2021-08-12 01:13: update #96
- 2021-08-12 22:25: update #97
- 2021-08-16 00:49: update #98
- 2021-08-16 00:18: update #99
- 2021-08-18 03:21: update #100
- 2021-08-20 01:28: update #101
- 2021-08-23 03:45: update #102
- 2021-08-24 03:47: update #103
- 2021-08-25 05:50: update #104
- 2021-08-27 00:12: update #105
- 2021-08-28 00:43: update #106
- 2021-08-30 19:11: update #107
- 2021-08-31 23:54: update #108
- 2021-09-02 22:52: update #109
- 2021-09-03 22:45: update #110
- 2021-09-05 05:19: update #111
- 2021-09-07 04:17: update #112
- 2021-09-07 18:43: update #113
- 2021-09-08 00:15: update #114
- 2021-09-08 18:40: update #115
- 2021-09-09 19:44: update #116
- 2021-09-11 01:50: update #117
- 2021-09-13 06:04: update #118
- 2021-09-14 00:50: update #119
- 2021-09-13 22:48: update #120
- 2021-09-15 05:47: update #121
- 2021-09-15 23:09: update #122
- 2021-09-16 18:28: update #123
- 2021-09-16 20:45: update #124
- 2021-09-20 04:41: update #125
- 2021-09-20 22:03: update #126
- 2021-09-22 02:22: update #127
- 2021-09-24 19:38: update #128
- 2021-09-25 18:03: update #129
- 2021-09-27 05:26: update #130
- 2021-09-28 02:52: update #131
- 2021-09-28 21:14: update #132
- 2021-09-28 17:08: update #133
- 2021-09-30 05:16: update #134
- 2021-10-02 01:29: update #135
- 2021-10-01 18:49: update #136
- 2021-10-04 03:36: update #137
- 2021-10-04 16:20: update #138
- 2021-10-05 17:02: update #139
- 2021-10-10 01:39: update #140
- 2021-10-10 17:02: update #141
- 2021-10-11 18:39: update #142
- 2021-10-12 23:16: update #143
- 2021-10-15 03:18: update #144
- 2021-10-15 05:41: update #145
- 2021-10-14 21:42: update #146
- 2021-10-16 06:05: update #147
- 2021-10-18 05:05: update #148
- 2021-10-18 01:46: update #149
- 2021-10-18 16:28: update #150
- 2021-10-20 03:43: update #151
- 2021-10-20 05:04: update #152
- 2021-10-20 17:09: update #153
- 2021-10-21 18:25: update #154
- 2021-10-27 01:39: update #155
- 2021-10-29 17:47: update #156
- 2021-11-01 02:41: update #157
- 2021-11-03 06:18: update #158
- 2021-11-05 03:20: update #159
- 2021-11-07 22:07: update #160
- 2021-11-10 19:26: update #161
- 2021-11-11 16:23: update #162
- 2021-11-12 18:56: update #163
- 2021-11-12 21:19: update #164
- 2021-11-16 16:36: update #165
- 2021-11-17 21:06: update #166
- 2021-11-19 04:46: update #167
- 2021-11-23 04:18: update #168
- 2021-11-22 17:29: update #169
- 2021-11-22 23:00: update #170
- 2021-11-23 21:13: update #171
- 2021-11-23 17:00: update #172
- 2021-11-24 19:27: update #173
- 2021-11-24 17:27: update #174
- 2021-11-26 03:41: update #175
- 2021-11-26 23:59: update #176
- 2021-11-29 20:36: update #177
- 2021-11-30 23:05: update #178
- 2021-12-03 06:27: update #179
- 2021-12-08 21:59: update #180
- 2021-12-09 22:14: update #181
- 2021-12-11 05:39: update #182
- 2021-12-13 16:29: update #183
- 2021-12-14 00:40: update #184
- 2021-12-14 05:49: update #185
- 2021-12-15 00:06: update #186
- 2021-12-14 22:35: update #187
- 2021-12-16 01:07: update #188
- 2021-12-18 00:59: update #189
- 2021-12-18 03:22: update #190

src/utils/color.ts
import { ColorValue } from './types';

/** Parse a hex color string to [r, g, b, a] */
export function parseColor(color: ColorValue): [number, number, number, number] {
  if (Array.isArray(color)) {
    if (color.length === 4) return color as [number, number, number, number];
    return [...color, 255] as [number, number, number, number];
  }

  const hex = color.replace('#', '');
  if (hex.length === 3) {
    const r = parseInt(hex[0] + hex[0], 16);
    const g = parseInt(hex[1] + hex[1], 16);
    const b = parseInt(hex[2] + hex[2], 16);
    return [r, g, b, 255];
  }

  if (hex.length === 6) {
    const r = parseInt(hex.slice(0, 2), 16);
    const g = parseInt(hex.slice(2, 4), 16);
    const b = parseInt(hex.slice(4, 6), 16);
    return [r, g, b, 255];
  }

  if (hex.length === 8) {
    const r = parseInt(hex.slice(0, 2), 16);
    const g = parseInt(hex.slice(2, 4), 16);
    const b = parseInt(hex.slice(4, 6), 16);
    const a = parseInt(hex.slice(6, 8), 16);
    return [r, g, b, a];
  }

  return [0, 0, 0, 255];
}

/** Convert [r, g, b, a?] to a hex string */
export function colorToHex(color: [number, number, number] | [number, number, number, number]): string {
  const r = Math.round(color[0]).toString(16).padStart(2, '0');
  const g = Math.round(color[1]).toString(16).padStart(2, '0');
  const b = Math.round(color[2]).toString(16).padStart(2, '0');
  return `#${r}${g}${b}`;
}

/** Linearly interpolate between two RGBA colors */
export function interpolateColor(
  from: [number, number, number, number],
  to: [number, number, number, number],
  t: number
): [number, number, number, number] {
  const ct = Math.max(0, Math.min(1, t));
  return [
    from[0] + (to[0] - from[0]) * ct,
    from[1] + (to[1] - from[1]) * ct,
    from[2] + (to[2] - from[2]) * ct,
    from[3] + (to[3] - from[3]) * ct,
  ];
}

/** A built-in set of color scale presets */
const COLOR_SCALE_PRESETS = {
  heat: [
    [0, 0, 128, 255],
    [0, 0, 255, 255],
    [0, 255, 255, 255],
    [0, 255, 0, 255],
    [255, 255, 0, 255],
    [255, 128, 0, 255],
    [255, 0, 0, 255],
  ],
  viridis: [
    [68, 1, 84, 255],
    [72, 40, 120, 255],
    [62, 74, 137, 255],
    [49, 104, 142, 255],
    [38, 130, 142, 255],
    [31, 158, 137, 255],
    [53, 183, 121, 255],
    [110, 206, 88, 255],
    [181, 222, 43, 255],
    [253, 231, 37, 255],
  ],
  plasma: [
    [13, 8, 135, 255],
    [75, 3, 161, 255],
    [125, 3, 168, 255],
    [168, 34, 150, 255],
    [203, 70, 121, 255],
    [229, 107, 93, 255],
    [248, 148, 65, 255],
    [253, 195, 40, 255],
    [240, 249, 33, 255],
  ],
  coolwarm: [
    [59, 76, 192, 255],
    [98, 130, 234, 255],
    [141, 176, 254, 255],
    [184, 208, 249, 255],
    [221, 221, 221, 255],
    [245, 196, 173, 255],
    [244, 154, 123, 255],
    [222, 96, 77, 255],
    [180, 4, 38, 255],
  ],
} as const;

export type ColorScalePreset = keyof typeof COLOR_SCALE_PRESETS;

/**
 * Create a color scale function from a preset or custom color stops.
 * Returns a function that maps a value in [0, 1] to an RGBA color.
 */
export function createColorScale(
  scaleOrStops: ColorScalePreset | [number, number, number, number][]
): (t: number) => [number, number, number, number] {
  const stops: [number, number, number, number][] =
    typeof scaleOrStops === 'string'
      ? (COLOR_SCALE_PRESETS[scaleOrStops] as unknown as [number, number, number, number][])
      : scaleOrStops;

  return (t: number): [number, number, number, number] => {
    const ct = Math.max(0, Math.min(1, t));
    const scaledIndex = ct * (stops.length - 1);
    const lowerIndex = Math.floor(scaledIndex);
    const upperIndex = Math.min(lowerIndex + 1, stops.length - 1);
    const localT = scaledIndex - lowerIndex;

    return interpolateColor(stops[lowerIndex], stops[upperIndex], localT);
  };
}

/**
 * Generate an array of evenly-spaced colors from a scale.
 */
export function sampleColorScale(
  scale: (t: number) => [number, number, number, number],
  count: number
): [number, number, number, number][] {
  const result: [number, number, number, number][] = [];
  for (let i = 0; i < count; i++) {
    result.push(scale(count === 1 ? 0.5 : i / (count - 1)));
  }
  return result;
}

/**
 * Compute a color for a value within a numeric domain using a color scale.
 */
export function domainColor(
  value: number,
  domain: [number, number],
  scale: (t: number) => [number, number, number, number]
): [number, number, number, number] {
  const [min, max] = domain;
  if (max === min) return scale(0.5);
  const t = (value - min) / (max - min);
  return scale(t);
}
